### DISCUSSION TITLE
Fortran Interface on initializing  with calculated flow field with multifab.

### DISCUSSION THREAD

**Discussion by paradoxknight1** [AMReX-Codes/amrex#4191] [Q&A] [answered]
Dear Amrex platform developer,

I'm quite newbie on Amrex, and using Fortran-interface to undertake my calculation.
My question could be seperated into two steps.
For the first one, I've almost gone through all of the case files for Fortran interface, but I have not found there is a checkpoint file writer interface for Fortran. May I check if writing the Fortran interface for checkpoint write is the only way for me to continue my calculation from previous data?

For the second one, I am facing a problem that asked me to "patch" part of the data from previously calculated results instead of the whole fields, and that data patching process needed me to place that data patch on other positons instead of its original place.

A good example to explain my question is the shockwave. After a period of time of calculation, the shockwave traveled from the initial side to the end side. After that, I need to take a patch of data that covers the shockwave's leeward and upwind side. And place it at the beginning side to continue the calculation. 

Is there any advice on my question?

Looking forward to hearing from you.

Many thanks.

**Comment by WeiqunZhang** [comment]
Different applications need different data for their checkpoints. So you need to write your functions. Nevertheless, you can use `amrex_multifab_write` and `amrex_multifab_read` to write and read binary MultiFab data. For many applications, the other data in checkpoint files are just some clear text files that are written out by the I/O process.

Re: "patch", is that just periodic boundary?

**Reply by paradoxknight1** [reply]
Dear Weiqun,
thanks for your timely reply.

For the "patch" process, it is more likely a trick to limit the calculation domain. For my problem, the shock wave needs to go through a nozzle first. And then, impainge a droplet. if the whole process is being calculated all at once, the RAM and time consumption will be extremely high. Therefore, the usual way to handle this issue is to calculate the shockwave interaction with nozzle first, and then patch the result to the droplet.

Therefore, a regular periodic boundary is not working for my cases, since I only need part of the data to "patch" it to the initial side, and at the same time, to initial a droplet at the "patched" shockwave front. 

And the whole process is more likely to use a _amrex_multifab_read_, to continue the calculation but needs to modify it first.

Since the dataptr() is only works with the mfi(mpi process), and it only works within each mpi_rank. I have no idea on how to move all the data(the box array, refinement information, mesh data, etc.) from one side to the other. As shown in the attached figure, the black box's data wants to be moved to the red box position. and fill the black box with initial field information.

![image](https://github.com/user-attachments/assets/6f314b42-5ff7-4041-81ab-523f6d6f5567)


Looking forwards to hearing from you.

Many thanks.

**Comment by nataraj2** [comment]
@paradoxknight1 
Here is a possible way for the issue. @WeiqunZhang will let you know if there is anything better or if there are any issues with this.

1. Here is an [example of the fortran advection routine with restart ](https://github.com/nataraj2/amrex-tutorials/tree/advection_fortan_restart/ExampleCodes/FortranInterface/Advection_F/Exec/SingleVortex). Go through the README section in the page for details.

2. For copying the data as you described you can do the following.

[This routine](https://github.com/AMReX-Codes/amrex/blob/development/Tools/Postprocessing/C_Src/PtwisePltTransform.cpp) gives a way of reading in the data in a plotfile and writing out another plotfile with the modified data. Hence, you can use this routine to do the data manipulation you need. Incorporate the [Restart routine](https://github.com/nataraj2/amrex-tutorials/blob/advection_fortan_restart/ExampleCodes/FortranInterface/Advection_F/Source/Restart.cpp) with this routine so that a [checkpoint file can be written as in here
](https://github.com/nataraj2/amrex-tutorials/blob/advection_fortan_restart/ExampleCodes/FortranInterface/Advection_F/Source/evolve_mod.F90#L72). Once a checkpoint file is written, you can restart the simulation with the modified data.

For the data manipulation, you can do as follows.
1. Run the code and get the plotfiles.
2. Read in the plotfile you want using[ the routine](https://github.com/AMReX-Codes/amrex/blob/development/Tools/Postprocessing/C_Src/PtwisePltTransform.cpp) above.
3. Create a multifab with only 1 level. The idea is that the level 0 data in the plotfile will be copied on to this new multifab. So, this step is only doing things with level 0. Use the `i` indices to do the data manipulation as follows

- Find the i indices to copy from - say i_copyfrom_1 to i_copyfrom_2. 
- Find the i indices to copy to - say i_copyto_1 to i_copyto_2.
- Loop over the level0 multifab data in the plotfile 2 times.  
- In the first for loop pass, store the data that you need to copy from (using the i indices) in an intermediate data object -- vector or array (or anything of your choice)
- In the second for loop pass, copy that data from the intermediate object to the indices that you want to copy it to in the new multifab that was created.

4. Write the plotfile with the modified data (to verify things are correct) and also the checkpoint file of the modified data.
6. Now you can restart the simulation with the checkpoint file. While restarting, you can specify the `amr.max_level` variable in the `inputs` to what you want. The code will automatically fill the finer levels.

**Reply by WeiqunZhang** [reply]
For copying data from one MultiFab to another with index shifting, there is a ParallelCopy function for it. For example,
```
        MultiBlockIndexMapping dtos{};                                                                                                                                       
        dtos.offset = IntVect(-100,0,0); // Copy data from (i,j,k) in srcmf to (i-100,j,k) in destmf
        ParallelCopy(destmf, geom.Domain(), srcmf, 0, 0, ncomp, IntVect(0), dtos);
```

**Reply by paradoxknight1** [reply]
Sorry for the late reply.
May I ask, what will happen for those data of mesh in destmf if the i-100 is less than the smallest end of the domain.

**Reply by WeiqunZhang** [reply]
If i-100 is outside the box specified by the second argument of ParallelCopy, nothing happens. If i-100 is inside the box, but outside destmf's index space, it's a segfault bug.

**Reply by paradoxknight1** [reply]
Thanks for your timely reply.
That is quite clear for me to understand. The parallel copy will only trying to transfer the geom.Domain() (a box object)'s data from the srcMF to destMF with a dtos shifting. Therefore, I have to make sure that the destination multifab should have the position to place the transfered data.

But I still wonder when index shifting a refined level of mesh, both the srcmf and destmf's index space are not continued, Let's assume that the multifabs only have defination at range of 1000-1200 at the begnining( if I use the same multifab as source and destination). Will the ParallelCopy function automaticlly transfer the boxarray from 1000-1200 to 900-1100. Though 900-1100 don't have defination at beginning.

Sorry for keep asking.

**Reply by WeiqunZhang** [reply]
Just to be clear, the Box argument is for the destination domain, not the source domain.
```
//! \param[in]  destbox The index box in the destination space that will be filled by data from                                                                                                                 
//!                     src. The source box that describes the dependencies will be computed by                                                                                                                 
//!                     the specified DTOS.                     
```

As for your question, I don't quite understand it. Maybe you can try to clarify it. Or maybe you can provide more context of what you want to do. Sometimes it might be an xy problem. https://en.wikipedia.org/wiki/XY_problem

**Reply by nataraj2** [reply, marked as answer]
@paradoxknight1 
Regarding your question of index shifting a refined level of mesh.

As mentioned in the step 3 in the reply on Oct18 above, the index shifting and copying of data is done only for the coarsest level mesh. So, there is no data copying involved for the other refinement levels. The idea is to do the copying only for level 0, write a checkpoint file for just the level 0 and use it to do the restart for the simulation. Note that the level 0 mesh remains the same at all times. While restarting the simulation, just use the number of refinement levels you want to use in the inputs file. The code will automatically fill all the refinement levels.

**Reply by WeiqunZhang** [reply]
> If i-100 is outside the box specified by the second argument of ParallelCopy, nothing happens. If i-100 is inside the box, but outside destmf's index space, it's a segfault bug.

I realized I was wrong about segfault. Nothing would happen. The function behaves like this
```
for each (i,j,k) in source multifab
    if (shifted(i,j,k) is inside destbox and shifted(i,j,k) is inside destination multifab
        dest(shifted(i,j,k)) = source(i,j,k)
```
I hope this makes things clearer.

**Reply by paradoxknight1** [reply]
For the  coarsest level we did the exact as your recommended @nataraj2, thanks you so much for your kindly help.
And for the finer level of mesh, we use the BoxArray's build-in function .shift( ), to build an temperoal mesh and then transfer the data into the newly built temperoal mesh.
code is something like.

`    for(int lev = 1; lev <= finest_level; ++lev ) {
        shifting = shifting * refRatio[lev-1];
        amrex::Print() << "level is" << lev << "and the indices shift is"<< shifting<<"\n";
        dtos.offset = shifting;
        MultiFab& phi_ori = *phi[lev]; // input multifab
        bat = BoxArray(*ba[lev]);
        amrex::Print() << bat <<"is original" <<"\n";
        bat.shift(shifting);
        amrex::Print() << bat <<"is shifted" <<"\n";
        geom_t = *geom[lev];
        domain1 = geom_t.Domain();
        DistributionMapping dm1 { bat, ParallelDescriptor::NProcs() };
        phi_temp[lev] = MultiFab(bat,dm1,Ncomp,Ngrow);
        amrex::Print() << "finished define a multifab"<< shifting<<"\n";
        ParallelCopy(phi_temp[lev],domain1.shift(shifting),phi_ori,0,0,Ncomp,amrex::IntVect(0),dtos);
    }`

It seems work perfectly right now. Thanks for everyone's help @WeiqunZhang @nataraj2
